lib/engine/hex.rb

```diff
@@ -181,32 +182,40 @@ def lay(tile)
 
       @tile = tile
 
-      @connections.clear
+      invalidate_connections
+      invalidate_connected
       @paths = nil
-
-      connect!
     end
 
-    def lay_downgrade(tile)
+    def invalidate_connected
       hexes = []
-
-      @tile.paths.each do |path|
-        path.walk { |p| hexes << p.hex if p.node? }
+      tile.paths.each do |path|
+        path.walk do |p|
+          hexes << p.hex
+        end
       end
 
-      lay(tile)
+      hexes.uniq.each(&:invalidate_connections)
+    end
 
-      hexes.uniq.each do |hex|
-        hex.connections.each do |_, connections|
-          connections.select!(&:valid?)
-        end
-      end
+    def invalidate_connections
+      @connected = false
+      @connections.clear
+    end
 
-      tile.restore_borders
+    def connections
+      connect unless @connected
+      @connections
     end
 
-    def connect!
-      Connection.connect!(self)
+    def connections_cached
```

Comment:
do we need this if we already have line 206 connections?

Reply:
i see, you're using it as a setter, i think that pattern is a bit weird

Reply:
perhaps it makes sense to move the connect code in Hex so you don't have to expose the private connections_cached, or just have the connection code set it directly

Reply:
It can call connections directly, will update